synapse/handlers/federation_event.py

```diff
@@ -895,12 +895,63 @@ async def _process_pulled_events(
             str(len(new_events)),
         )
 
-        # We want to sort these by depth so we process them and
-        # tell clients about them in order.
-        sorted_events = sorted(new_events, key=lambda x: x.depth)
-        for ev in sorted_events:
-            with nested_logging_context(ev.event_id):
-                await self._process_pulled_event(origin, ev, backfilled=backfilled)
+        @trace
+        async def _process_new_pulled_events(new_events: Collection[EventBase]) -> None:
+            # We want to sort these by depth so we process them and
+            # tell clients about them in order.
+            sorted_events = sorted(new_events, key=lambda x: x.depth)
+            for ev in sorted_events:
+                with nested_logging_context(ev.event_id):
+                    await self._process_pulled_event(origin, ev, backfilled=backfilled)
+
+        # Check if we've already tried to process these events at some point in the
+        # past. We aren't concerned with the expontntial backoff here, just whether it
+        # has failed to be processed before.
+        event_ids_with_failed_pull_attempts = (
+            await self._store.get_event_ids_with_failed_pull_attempts(
+                [event.event_id for event in new_events]
+            )
+        )
```

Comment:
Should this be a `set()` to avoid the potential `O(n^2)` that happens twice below? (I don't know how long these lists are and if it matters...)

Reply:
We could easily be more efficient here if we drop the source order constraint from MSC2716 (mentioned in the [other comment](https://github.com/matrix-org/synapse/pull/15585#discussion_r1204514303)).

Since my plan is to drop MSC2716 anyway, perhaps we could do some of that work first. Sucks to have this blocked on that though.

Reply:
Discussed more in the [#synapse-dev room](https://matrix.to/#/!vcyiEtMVHIhWXcJAfl:sw1v.org/$VyVNUaLGyL48B3bZcimZyuR_WQukJFXPKsAvqE_yqhg?via=matrix.org&via=element.io&via=pixie.town) and thanks to @clokep for further poking this point so we can move this PR forward without more labor :tada: 

We can totally update this to be a `set()` (for efficient lookups) without affecting anything.

And for some reason, I was thinking there was an additional piece to do with the other list comprehensions but they're just good and nothing needs to change there